# 异常断电后celerybeat无法启动

## 问题描述

* 在客户的某个项目中, 我们使用celery来作为分布式任务系统; 某次异常断电后, 定时任务出现问题

* 经过诊断发现celerybeat服务无法启动, 一般来说一个celery系统中只能存在一个beat进程, 所以beat进程默认会在工作目录下创建一个celerybeat.pid来记录其进程PID, 避免运行多个beat进程

* 当beat被停止时, celerybeat.pid文件被会自动删除; 当启动beat时, beat会检测celerybeat.pid文件是否存在, 如果存在还会检测其中记录的pid对应的进程是否存在, 如果存在, 就不会启动; 反之则启动

* beat有个缺陷, 就是不会判断pid对应的进程是否为相应的beat进程, 而仅仅只是判断pid对应的进程是否存在

## 问题原因

* 异常断电时, beat无法及时删除celerybeat.pid文件, 导致该文件残留

* 启动系统后, 恰好存在某个进程的PID与残留的celerybeat.pid文件中记录的PID一致, 导致beat服务无法启动

## 问题复现

* 停止beat服务, 在beat服务的工作目录创建celerybeat.pid文件, 并在里面写入一个任意进程的PID

* 启动beat服务, 发现无法启动

## 解决办法

方法1: 修改celery源码, 在celerybeat.pid中记录本次运行的二进制文件和参数, 在启动beat时, 检测记录的PID是否存在, 如果存在, 还要判断PID对应的进程的可执行文件和参数是否与记录文件中的一致

方法2: **方法1过于复杂了, 而且检测进程是否为beat进程的方法也存在缺陷, 只是一个思路而已**

    在beat的启动参数中设置--pidfile=/run/celerybeat.pid, /run是一个临时目录, 在异常断电后被自动删除, 所以不会导致pid文件残留, 也就不会导致上述问题发生了

方法3: 禁用celery的pid文件, 使用upstart或者systemd的pid文件

    在beat的启动参数中添加--pidfile=来禁用掉自带的pid文件